rust/monovertex/src/forwarder.rs

```diff
@@ -106,21 +107,61 @@ impl Forwarder {
                     // TODO: should we retry writing? what if the error is transient?
                     //    we could rely on gRPC retries and say that any error that is bubbled up is worthy of non-0 exit.
                     //    we need to confirm this via FMEA tests.
-                    let start_time = tokio::time::Instant::now();
-                    self.sink_client.sink_fn(transformed_messages).await?;
-                    info!("Sink latency - {}ms", start_time.elapsed().as_millis());
-                    counter!(FORWARDER_WRITE_TOTAL, &self.common_labels).increment(messages_count);
+                    let mut retry_messages = transformed_messages;
+                    let mut attempts = 0;
+                    let mut error_map = HashMap::new();
 
-                    // Acknowledge the messages
-                    // TODO: should we retry acking? what if the error is transient?
-                    //    we could rely on gRPC retries and say that any error that is bubbled up is worthy of non-0 exit.
-                    //    we need to confirm this via FMEA tests.
-                    let start_time = tokio::time::Instant::now();
-                    self.source_client.ack_fn(offsets).await?;
-                    info!("Ack latency - {}ms", start_time.elapsed().as_millis());
+                    while attempts <= config().sink_max_retry_attempts {
+                        let start_time = tokio::time::Instant::now();
+                        match self.sink_client.sink_fn(retry_messages.clone()).await {
+                            Ok(response) => {
+                                info!("Sink latency - {}ms", start_time.elapsed().as_millis());
+
+                                 let failed_ids: Vec<String> = response.results.iter()
+                                    .filter(|result| result.status != proto::Status::Success as i32)
+                                    .map(|result| result.id.clone())
+                                    .collect();
+
+                                let successful_offsets: Vec<Offset> = retry_messages.iter()
+                                    .filter(|msg| !failed_ids.contains(&msg.id))
+                                    .map(|msg| msg.offset.clone())
+                                    .collect();
+
+
+                                // ack the successful offsets
+                                let n = successful_offsets.len();
+                                self.source_client.ack_fn(successful_offsets).await?;
```

Comment:
we cannot do this, in Kafka if we ack `n`th message, everything upto `n` is considered as processing. If the pod crashes, Kafka will not redeliver those messages that we are still retrying.